feat(aws-codebuild): allow github sourceversion branch

knorms101 opened the pull request:
Addresses issue #5777.

Allows users to specify which GitHub branch a CodeBuild should use as its Source Version. Currently the branch always defaults to 'master', but users may want to test features on other  branches



----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

<!-- 
Please read the contribution guidelines and follow the pull-request checklist:
https://github.com/aws/aws-cdk/blob/master/CONTRIBUTING.md
 -->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* Result: FAILED
* [Build Logs](https://dfq49x5zyi.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=44fe1a09-e48f-4448-b2f5-521206f4c419%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* Result: SUCCEEDED
* [Build Logs](https://dfq49x5zyi.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=4bad8267-9646-4647-867a-a2985b24872f%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* Result: SUCCEEDED
* [Build Logs](https://dfq49x5zyi.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=c73bd72c-65bb-43b7-a899-39e952fc6a2f%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

skinny85 commented on packages/@aws-cdk/aws-codebuild/lib/source.ts:
```diff
@@ -19,6 +19,13 @@ export interface SourceConfig {
+   * @see http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-codebuild-project.html#cfn-codebuild-project-sourceversion
+   * @default the latest version
+   */
+  readonly sourceVersion?: string | undefined;
```
Just `sourceVersion?: string` is enough to be optional, no need for the `| undefined` 🙂.

skinny85 commented on packages/@aws-cdk/aws-codebuild/lib/source.ts:
```diff
@@ -554,18 +561,28 @@ export interface GitHubSourceProps extends ThirdPartyGitSourceProps {
  */
 class GitHubSource extends ThirdPartyGitSource {
   public readonly type = GITHUB_SOURCE_TYPE;
+  public readonly branch?: string;
```
Why `public`?

skinny85 commented on packages/@aws-cdk/aws-codebuild/test/test.codebuild.ts:
```diff
@@ -479,6 +479,7 @@ export = {
         source: codebuild.Source.gitHub({
           owner: 'testowner',
           repo: 'testrepo',
+          branch: 'testbranch',
```
I think the test you added below is fine - no need for this one :).

knorms101 replied:
👍 Sounds good. I removed this test.

skinny85 commented on packages/@aws-cdk/aws-codebuild/lib/source.ts:
```diff
@@ -554,18 +561,28 @@ export interface GitHubSourceProps extends ThirdPartyGitSourceProps {
   readonly repo: string;
+
+  /**
+   * The name of the branch.
```
Can you also specify that you can give tags, commit IDs, and PR IDs, according to http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-codebuild-project.html#cfn-codebuild-project-sourceversion ?

knorms101 replied:
Makes sense. I renamed the property so "sourceVersion" to be more general and updated the doc to reflect the variety of options.

knorms101 commented:
> Thanks for the contribution @knorms101 , it's great! I have a few small comments.
> 
> From looking at the docs of this feature in http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-codebuild-project.html#cfn-codebuild-project-sourceversion , I see this also works for CodeCommit, BitBucket and S3. Do you have time to add those as well? (S3 probably should be called `version` instead of `branch`).
> 
> Thanks!

Thanks for the feedback, @skinny85! I agree that it would be great to include this option for the other CodeBuild sources too. I'll work on adding that next week.

skinny85 commented:
> > Thanks for the contribution @knorms101 , it's great! I have a few small comments.
> > From looking at the docs of this feature in http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-codebuild-project.html#cfn-codebuild-project-sourceversion , I see this also works for CodeCommit, BitBucket and S3. Do you have time to add those as well? (S3 probably should be called `version` instead of `branch`).
> > Thanks!
> 
> Thanks for the feedback, @skinny85! I agree that it would be great to include this option for the other CodeBuild sources too. I'll work on adding that next week.

Awesome, thanks so much for taking the time to work on this! ❤️

knorms101 commented:
> > > Thanks for the contribution @knorms101 , it's great! I have a few small comments.
> > > From looking at the docs of this feature in http://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-codebuild-project.html#cfn-codebuild-project-sourceversion , I see this also works for CodeCommit, BitBucket and S3. Do you have time to add those as well? (S3 probably should be called `version` instead of `branch`).
> > > Thanks!
> > 
> > 
> > Thanks for the feedback, @skinny85! I agree that it would be great to include this option for the other CodeBuild sources too. I'll work on adding that next week.
> 
> Awesome, thanks so much for taking the time to work on this! ❤️

Hi @skinny85 I finally had a chance to continue working on this. I added the optional `sourceVersion` property to all the other CodeBuild source options. I also added the optional `secondarySourceVersions` property.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* Result: SUCCEEDED
* [Build Logs](https://dfq49x5zyi.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=525c3b6b-9c55-41a2-8fea-201921f683d4%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

skinny85 commented on packages/@aws-cdk/aws-codebuild/lib/project.ts:
```diff
@@ -537,6 +537,13 @@ export interface ProjectProps extends CommonProjectProps {
+   * @default - No secondary source versions (they default to latest).
+   * @see https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-codebuild-project.html#cfn-codebuild-project-sourceversion
+   */
+  readonly secondarySourceVersions?: CfnProject.ProjectSourceVersionProperty[];
```
This should not be here, right? It's not actually used anywhere.

Every `ISource` passed in `secondarySources` has their own `sourceVersion` property returned from `bind`, and that's what should be filled for the `secondarySourceVersions` property of `CfnProject`.

Also, remember that we never use L1 types (those that start with `Cfn*`) for L2 properties.

Please remove this property.

skinny85 commented on packages/@aws-cdk/aws-codebuild/lib/project.ts:
```diff
@@ -626,7 +633,7 @@ export class Project extends ProjectBase {
    */
   public static serializeEnvVariables(environmentVariables: { [name: string]: BuildEnvironmentVariable }):
-      CfnProject.EnvironmentVariableProperty[] {
+    CfnProject.EnvironmentVariableProperty[] {
```
Please leave this indentation as it is.

skinny85 commented on packages/@aws-cdk/aws-codebuild/lib/project.ts:
```diff
@@ -654,6 +661,7 @@ export class Project extends ProjectBase {
   private readonly source: ISource;
   private readonly buildImage: IBuildImage;
   private readonly _secondarySources: CfnProject.SourceProperty[];
+  private readonly _secondarySourceVersions: CfnProject.ProjectSourceVersionProperty[];
```
This should be called '_secondarySource**s**Versions', to be consistent with `_secondarySources`.

knorms101 replied:
The [CloudFormation property](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-codebuild-project.html#cfn-codebuild-project-secondarysourceversions) is called "SecondarySourceVersions." Wouldn't it be better to be consistent with the CF names?

skinny85 commented on packages/@aws-cdk/aws-codebuild/lib/project.ts:
```diff
@@ -788,7 +804,7 @@ export class Project extends ProjectBase {
       const projectStack = Stack.of(this);
       if (!(options.artifactBucket.encryptionKey instanceof kms.Key &&
-          (keyStack.account !== projectStack.account || keyStack.region !== projectStack.region))) {
+        (keyStack.account !== projectStack.account || keyStack.region !== projectStack.region))) {
```
Please leave this indentation as it is.

skinny85 commented on packages/@aws-cdk/aws-codebuild/lib/project.ts:
```diff
@@ -893,6 +909,12 @@ export class Project extends ProjectBase {
 
+  private renderSecondarySourceVersions(): CfnProject.ProjectSourceVersionProperty[] | undefined {
+    return this._secondarySourceVersions.length === 0
+    ? undefined
```
Indentation is off here, should be:

```ts
  private renderSecondarySourceVersions(): CfnProject.ProjectSourceVersionProperty[] | undefined {
    return this._secondarySourceVersions.length === 0
      ? undefined
      : this._secondarySourceVersions;
  }
```

skinny85 commented on packages/@aws-cdk/aws-codebuild/lib/project.ts:
```diff
@@ -983,8 +1005,8 @@ export class Project extends ProjectBase {
     if ((sourceType === CODEPIPELINE_SOURCE_ARTIFACTS_TYPE ||
-        artifactsType === CODEPIPELINE_SOURCE_ARTIFACTS_TYPE) &&
-        (sourceType !== artifactsType)) {
+      artifactsType === CODEPIPELINE_SOURCE_ARTIFACTS_TYPE) &&
```
Please leave the indentation as it is here.

skinny85 commented on packages/@aws-cdk/aws-codebuild/lib/project.ts:
```diff
@@ -1124,7 +1146,7 @@ class ArmBuildImage implements IBuildImage {
     const ret = [];
     if (buildEnvironment.computeType &&
-        buildEnvironment.computeType !== ComputeType.LARGE) {
+      buildEnvironment.computeType !== ComputeType.LARGE) {
```
Please leave the indentation as it is here.

skinny85 commented on packages/@aws-cdk/aws-codebuild/lib/source.ts:
```diff
@@ -471,6 +478,14 @@ abstract class ThirdPartyGitSource extends GitSource {
+   * @example 'mybranch'
+   * @default the default branch's HEAD commit ID is used
+   */
+  readonly sourceVersion?: string;
```
I actually don't like this name ;(. In my opinion, 99% of the time, this will be a branch.

What do you think about the name `branchOrRef`?

knorms101 replied:
👍  I like that name.

skinny85 commented on packages/@aws-cdk/aws-codebuild/lib/source.ts:
```diff
@@ -508,6 +526,13 @@ class CodeCommitSource extends GitSource {
+   *
+   * @default latest
+   */
+  readonly sourceVersion?: string;
```
This should be just called `version` in my opinion (it references an S3 version).

skinny85 commented on packages/@aws-cdk/aws-codebuild/lib/source.ts:
```diff
@@ -554,18 +582,29 @@ export interface GitHubSourceProps extends ThirdPartyGitSourceProps {
+   * @example 'pr/25'
+   * @default the default branch's HEAD commit ID is used
+   */
+  readonly sourceVersion?: string;
```
Same comment here (`branchOrRef`?).

skinny85 commented on packages/@aws-cdk/aws-codebuild/lib/source.ts:
```diff
@@ -554,18 +582,29 @@ export interface GitHubSourceProps extends ThirdPartyGitSourceProps {
  */
 class GitHubSource extends ThirdPartyGitSource {
   public readonly type = GITHUB_SOURCE_TYPE;
+  private readonly sourceVersion?: string;
```
This probably should be below `httpsCloneUrl`.

skinny85 commented on packages/@aws-cdk/aws-codebuild/lib/source.ts:
```diff
@@ -595,6 +635,15 @@ export interface GitHubEnterpriseSourceProps extends ThirdPartyGitSourceProps {
+   * @example 'pr/25'
+   * @default the default branch's HEAD commit ID is used
+   */
+  readonly sourceVersion?: string;
```
Are we 100% sure this is supported for `GitHubEnterprise` Sources?

skinny85 replied:
Actually, here's an idea.

If it is supported for GitHub Enterprise, then we have a common parent of all Git-backed sources in this file, called `GitSource`. You can just add `brachOrRef` as a property there, instead of duplicating this code in every subclass!

knorms101 replied:
[This page](https://docs.aws.amazon.com/codebuild/latest/userguide/sample-source-version.html) seems to confirm that it is supported with `GithubEnterprise` sources.

Love that idea. I added `branchOrRef` to `GitSource` and removed it from every subclass, as you suggested.

skinny85 commented on packages/@aws-cdk/aws-codebuild/lib/source.ts:
```diff
@@ -641,6 +693,14 @@ export interface BitBucketSourceProps extends ThirdPartyGitSourceProps {
+   * @example 'mybranch'
+   * @default the default branch's HEAD commit ID is used
+   */
+  readonly sourceVersion?: string;
```
Same comment here (branchOrRef?).

skinny85 commented on packages/@aws-cdk/aws-codebuild/test/integ.project-secondary-sources-artifacts.expected.json:
```diff
@@ -209,6 +209,11 @@
+        "SecondarySourceVersions": [
+          {
+            "SourceIdentifier": "AddSource1"
+          }
```
Hmm. This is a problem.

I think you should filter rendering on whether `sourceVersion` is !== `undefined`. For example, this property should not be rendered in this case.

knorms101 replied:
I have updated so `SecondarySourceVersions` won't render if no versions are given.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* Result: SUCCEEDED
* [Build Logs](https://dfq49x5zyi.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=2db18ecf-f5a6-4dbb-ab9a-6670017c7330%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* Result: SUCCEEDED
* [Build Logs](https://dfq49x5zyi.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=c245050c-2617-4fac-8608-80d22ff19983%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

skinny85 commented on packages/@aws-cdk/aws-codebuild/lib/source.ts:
```diff
@@ -479,10 +495,12 @@ export interface CodeCommitSourceProps extends GitSourceProps {
   constructor(props: CodeCommitSourceProps) {
     super(props);
     this.repo = props.repository;
+    this.branchOrRef = props.branchOrRef;
```
There is actually a way to get rid of this duplication.

In `GitSource`:

```ts
abstract class GitSource extends Source {
  private readonly cloneDepth?: number;
  private readonly branchOrRef?: string;

  protected constructor(props: GitSourceProps) {
    super(props);

    this.cloneDepth = props.cloneDepth;
    this.branchOrRef = props.branchOrRef;
  }

  public bind(_scope: Construct, _project: IProject): SourceConfig {
    const superConfig = super.bind(_scope, _project);
    return {
      ...superConfig,
      sourceVersion: this.branchOrRef,
      sourceProperty: {
        ...superConfig.sourceProperty,
        gitCloneDepth: this.cloneDepth,
      },
    };
  }
}
```

Then in every place that extends `GitSource` (I'm doing `CodeCommitSource`, but the other ones will be the same):

```ts
class CodeCommitSource extends GitSource {
  public readonly type = CODECOMMIT_SOURCE_TYPE;
  private readonly repo: codecommit.IRepository;

  constructor(props: CodeCommitSourceProps) {
    super(props);
    this.repo = props.repository;
  }

  public bind(_scope: Construct, project: IProject): SourceConfig {
    // https://docs.aws.amazon.com/codebuild/latest/userguide/setting-up.html
    project.addToRolePolicy(new iam.PolicyStatement({
      actions: ['codecommit:GitPull'],
      resources: [this.repo.repositoryArn],
    }));
    const superConfig = super.bind(_scope, project);
    return {
      ...superConfig,
      sourceProperty: {
        ...superConfig.sourceProperty,
        location: this.repo.repositoryCloneUrlHttp,
      },
    };
  }
}

knorms101 replied:
I removed the `branchOrRef` property from all the subclasses and moved it onto `GitSource` and `GitThirdPartySource`, then made the subclasses explicitly assign `sourceVersion: superConfig.sourceVersion`.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* Result: SUCCEEDED
* [Build Logs](https://dfq49x5zyi.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=a788a057-042c-40d9-a2ba-323e910d1b8b%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* Result: SUCCEEDED
* [Build Logs](https://dfq49x5zyi.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=6465e027-1601-49d2-bed6-3627fdc5b765%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*
